Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -33,14 +121,206 @@ theorem mem_automation_test_2
+    (h : read_mem_bytes 16 src_addr s = out) :
+    read_mem_bytes 10 (src_addr + 6) s = out.extractLsBytes 6 10 := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  · congr
+    -- ⊢ (src_addr + 6).toNat - src_addr.toNat = 6
+    bv_omega'
+/--
+info: 'ReadOverlappingRead.overlapping_read_test_2' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms overlapping_read_test_2
+
+/-- A read in the goal state overlaps with a read in the
+left hand side of the hypotheis `h`.
+-/
+theorem overlapping_read_test_3
+    (hlegal : mem_legal' src_addr 16)
+    (h : read_mem_bytes 16 src_addr s = read_mem_bytes 16 other_addr s) :
+    read_mem_bytes 10 (src_addr + 6) s =
+    -- @bollu: unfortunate, this needs `s.mem` to be public. Annoying.
+      (Memory.read_bytes 16 other_addr s.mem).extractLsBytes 6 10 := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  · congr
+    -- ⊢ (src_addr + 6).toNat - src_addr.toNat = 6
+    bv_omega'
+/--
+info: 'ReadOverlappingRead.overlapping_read_test_3' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms overlapping_read_test_3
+
+/- TODO(@bollu): This test case hangs at `bv_omega'`. This is to be debugged.
+/-- A read in the goal state overlaps with a read in the
+right hand side of the hypotheis `h`.
+-/
+theorem overlapping_read_test_4
+    (hlegal : mem_legal' src_addr 16)
+    (h : read_mem_bytes 16 other_addr s = read_mem_bytes 16 src_addr s) :
+    read_mem_bytes 10 (src_addr + 6) s =
+    -- @bollu: unfortunate, this needs `s.mem` to be public. Annoying.
+      (Memory.read_bytes 16 other_addr s.mem).extractLsBytes 6 10 := by
+  simp only [memory_rules] at h ⊢
+  simp_mem
+  · congr
+    -- ⊢ (src_addr + 6).toNat - src_addr.toNat = 6
+    bv_omega' -- TODO: Lean gets stuck here?
+
+#guard_msgs in #print axioms overlapping_read_test_4
+-/
+end ReadOverlappingRead
+
+namespace ReadOverlappingWrite
+
+theorem test_1 {val : BitVec (16 * 8)}
+    (hlegal : mem_legal' src_addr 16) :
+    Memory.read_bytes 16 src_addr (Memory.write_bytes 16 src_addr val mem) =
+     val.extractLsBytes 0 16  := by
+  simp_mem
+  · -- ⊢ val.extractLsBytes (src_addr.toNat - src_addr.toNat) 16 = val.extractLsBytes 0 16
+    congr
+    simp
+
+theorem test_2 {val : BitVec _}
+    (hlegal : mem_legal' src_addr 16) :
+    Memory.read_bytes 6 (src_addr + 10) (Memory.write_bytes 16 src_addr val mem) =
+    val.extractLsBytes 10 6 := by
+  simp_mem
+  have : ((src_addr + 10).toNat - src_addr.toNat) = 10 := by bv_omega'
+  rw [this]
+
+/--
+TODO(@bollu): the definition of overlap doesn't seem to do well with zero width!
+That's pretty interesting. I wonder if we ever need this though.
+-/
+theorem test_write_zero (hlegalw : mem_legal' write_addr 0)
+    (hlegalr : mem_legal' read_addr read_n) :
+    Memory.read_bytes read_n read_addr (Memory.write_bytes 0 write_addr write_val mem) =
+    mem.read_bytes read_n read_addr := by
+  simp_mem
+  sorry
```

Comment:
Perhaps we need a rule that `Memory.write_bytes 0 addr val mem` is just `mem`.

Reply:
@shigoel aha, that's a nice way to deal with this particular case. But more generally, I would like to improve the definition of `mem_subset` to allow zero width subsets.

Reply:
Added an issue to track, because in general, we may not know that the width is zero directly, but could be implied by some kind of complex `omega` condition. Tracking here: https://github.com/leanprover/LNSym/issues/116

Reply:
@bollu Adding a zero-byte `Memory.write_bytes`/`write_mem_bytes` theorem (and an analogous zero read one) seems like a good idea that's orthogonal to making `mem_subset` tolerate zero-width subsets.

Reply:
Supportive of making `mem_subset` more general, ofc!